### PR TITLE
Fix #319 print not working

### LINKED ISSUE #319: Printing failed to work for Pages and Slides mode
`viewer-container` is currently set to 1px by default.

However the print target is now `viewer-container` which will render everything into 1 px.

### PR DESCRIPTION
**Pull request by amoshydra**
Printing now work #319  and uses different technique.

Previously it rely solely on the CSS print media to hide and print. 
Current implement will make sure a full re-rendering of the content in Page Mode before on print.
It will intercept the CTRL/CMD + P command and execute the print command.

Since the print command will only intercept keyboard command, here are the known problem:
 - Clicking print via Browser's file context will not work.
 - Clicking print on mobile device will not work.

### CODE REVIEW

File: sashimi-webapp/static/styles/viewer-page.css [hunk 1]
```diff
@@ -21,3 +21,13 @@ body {
...
+
+@media print {
+  .page-view {
+    left: 0 !important;
+    margin: 0 auto !important;
+  }
+}
+@page {
+  margin: 0;
+}
```
Comment (tharain) [inline sashimi-webapp/static/styles/viewer-page.css]: eslint new line?

Comment (amoshydra) [reply sashimi-webapp/static/styles/viewer-page.css]: Fixed! eea3a23

File: sashimi-webapp/src/helpers/elementUtils.js [hunk 2]
```diff
@@ -10,6 +10,9 @@ export default {
     if (typeof targetReference === 'string') {
       const doc = altDocument || document;
       return doc.getElementById(targetReference);
+    } else if (targetReference.window === targetReference) {
```
Comment (chuajiaxuan) [inline sashimi-webapp/src/helpers/elementUtils.js:14]: How come an Element Object has a window properties? Is the jsdoc updated? Thank you!

File: sashimi-webapp/src/helpers/documentBuilder.js [hunk 3]
```diff
@@ -21,17 +21,33 @@ function constructStyleLink(renderDoc, link) {
 }
 
 export default {
-  getDocument(iframeElement) {
-    return iframeElement.contentWindow.document;
+  /**
+   * Get the document object of a iframe element
+   * @param {Element} An iframe element or a window object
```
Comment (chuajiaxuan) [inline sashimi-webapp/src/helpers/documentBuilder.js]: Should be {Element | Window} according to description.

File: sashimi-webapp/src/logic/inputHandler/DocumentPrinter/index.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import elementUtils from 'src/helpers/elementUtils';
+import core from './core';
+
+const keyEventListener = function keyEventListener(event) {
+  if (event.key === 'p' && (event.ctrlKey || event.metaKey)) {
+    // metaKey will match Window key and Mac's CMD key
```
Comment (chuajiaxuan) [inline sashimi-webapp/src/logic/inputHandler/DocumentPrinter/index.js:6]: Do you foresee any other issues with this part besides the window key on Windows? E.g. different browser compatibility? Cause I have been seeing this about the event.metaKey. (See first and second answers part)

http://stackoverflow.com/questions/3902635/how-does-one-capture-a-macs-command-key-via-javascript

**Review by chuajiaxuan — COMMENTED** [review]
Good job as always! Never fail to be wow-ed and learn from your codes! Besides that only the few comments below, otherwise good to go!! 👍

File: sashimi-webapp/src/logic/inputHandler/DocumentPrinter/index.js (hunk 4 above)
Comment (amoshydra) [reply sashimi-webapp/src/logic/inputHandler/DocumentPrinter/index.js:6]: It was a real issue a few years back before 2012.

However, on modern browser MetaKey is mostly assigned to CMD in Mac.
See https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/metaKey

Also, check the more recent answer in the stackoverflow link you provided http://stackoverflow.com/questions/3902635/how-does-one-capture-a-macs-command-key-via-javascript?answertab=active#comment43996640_3922353

File: sashimi-webapp/src/helpers/elementUtils.js (hunk 2 above)
Comment (amoshydra) [reply sashimi-webapp/src/helpers/elementUtils.js:14]: Fixed! Nice catch. 90420fe8

File: sashimi-webapp/src/logic/inputHandler/DocumentPrinter/index.js (hunk 4 above)
Comment (chuajiaxuan) [reply sashimi-webapp/src/logic/inputHandler/DocumentPrinter/index.js:6]: Ahhhh i see ok ok.

File: sashimi-webapp/src/helpers/documentBuilder.js (hunk 3 above)
Comment (amoshydra) [reply sashimi-webapp/src/helpers/documentBuilder.js]: Fixed! thanks 5b127b69

**Review by tharain — APPROVED** [review]
:100:

**Review by chuajiaxuan — APPROVED** [review]
👍 💯 👍

**Review by jiayingy — APPROVED** [review]

**Pull request merged**